goaci.go

```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
+// getBinaryName get a binary name built by go get and selected by
+// --use-binary parameter.
+func getBinaryName(opts *options, pathsNames *pathsAndNames) (string, error) {
+	fi, err := ioutil.ReadDir(pathsNames.goBinPath)
 	if err != nil {
-		die(err.Error())
+		return "", err
 	}
+
 	switch {
 	case len(fi) < 1:
-		die("no binaries found in gobin")
+		return "", fmt.Errorf("No binaries found in gobin.")
+	case len(fi) == 1:
+		name := fi[0].Name()
+		if opts.useBinary != "" && name != opts.useBinary {
+			return "", fmt.Errorf("No such binary found in gobin: %s. There is only %s", opts.useBinary, name)
+		}
+		Debug("found binary: ", name)
+		return name, nil
 	case len(fi) > 1:
-		debug(fmt.Sprint(fi))
-		die("can't handle multiple binaries")
+		names := []string{}
+		for _, v := range fi {
+			names = append(names, v.Name())
+		}
+		if opts.useBinary == "" {
+			return "", fmt.Errorf("Found multiple binaries in gobin, but --use-binary option is not used. Please specify which binary to put in ACI. Following binaries are available: %s", strings.Join(names, ", "))
+		}
+		for _, v := range names {
+			if v == opts.useBinary {
+				return v, nil
+			}
+		}
+		return "", fmt.Errorf("No such binary found in gobin: %s. There are following binaries available: %s", opts.useBinary, strings.Join(names, ", "))
+	}
+	return "", fmt.Errorf("Reaching this point shouldn't be possible.")
+}
+
+func getApp(opts *options, binary string) *types.App {
+	exec := []string{filepath.Join("/", binary)}
+	exec = append(exec, opts.exec...)
+
+	return &types.App{
+		Exec:  exec,
+		User:  "0",
+		Group: "0",
 	}
-	fn := fi[0].Name()
-	debug("found binary: ", fn)
+}
 
-	// Set up rootfs for ACI layout
-	rfs := filepath.Join(acidir, "rootfs")
-	err = os.MkdirAll(rfs, 0755)
+func getVCSLabel(pathsNames *pathsAndNames) (*types.Label, error) {
+	name, value, err := GetVCSInfo(pathsNames.projectPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get VCS info: %v", err)
+	}
+	acname, err := types.NewACName(name)
 	if err != nil {
-		die(err.Error())
+		return nil, fmt.Errorf("Invalid VCS label: %v", err)
 	}
+	return &types.Label{
+		Name:  *acname,
+		Value: value,
+	}, nil
+}
 
-	// Move the binary into the rootfs
-	ep := filepath.Join(rfs, fn)
-	err = os.Rename(filepath.Join(gobin, fn), ep)
+func prepareManifest(opts *options, pathsNames *pathsAndNames, binary string) (*schema.ImageManifest, error) {
+	name, err := types.NewACName(pathsNames.imageACName)
+	// TODO(jonboulle): could this ever actually happen?
+	if err != nil {
+		return nil, err
+	}
+
+	app := getApp(opts, binary)
+
+	vcsLabel, err := getVCSLabel(pathsNames)
 	if err != nil {
-		die(err.Error())
+		return nil, err
+	}
+	labels := types.Labels{
+		*vcsLabel,
+	}
+
+	manifest := schema.BlankImageManifest()
+	manifest.Name = *name
+	manifest.App = app
+	manifest.Labels = labels
+	return manifest, nil
+}
+
+func copyAssets(opts *options, pathsNames *pathsAndNames) error {
+	paths := map[string]string{
```

Comment:
Instead of ``paths``, I would name it ``placeholderMapping`` or something more descriptive.

Same thing in PrepareAssets() below.